File: cpp/0016_3SumClosest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(),nums.end());
        int offset=INT_MAX;
        int res=0;
        for(int i=0;i<nums.size()-2;i++)
        {
            int left=i+1;
            int right=nums.size()-1;
            while(left<right)
            {
                int sum=nums[i]+nums[left]+nums[right];
                int tmp= sum>target?sum-target:target-sum;
                if(tmp<offset)
                {
                    offset=tmp;
                    res=sum;
                }
                if(sum>target)
                {
                    right--;
                }
                else if(sum<target)
                {
                    left++;
                
                }
                else
                    return res;            }
        }
    return res;
        
    }
};

File: python/0016_3SumClosest.py

class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        res,offset=(nums[0]+nums[1]+nums[2]),abs(target-(nums[0]+nums[1]+nums[2]))
        nums.sort()#从小到大

        for i in range(len(nums)-2):
            # if i>0 and nums[i] ==nums[i-1]:
            #     continue

            left,right=i+1,len(nums)-1
            while right>left:
                isum=nums[i]+nums[left]+nums[right]
                tmp=abs(target-isum)
                if tmp<offset:
                    offset=tmp
                    res=isum

                if isum>target:
                    right -= 1
                elif isum<target:
                    left+=1
                else:
                    return res

        return res
        


